### PR TITLE
Additional reverse geo lookup provider and can run as docker container

### PR DESCRIPTION
**Pull request by halteren**

### CODE REVIEW

File: geo.c [hunk 1]
```diff
@@ -163,6 +250,9 @@ bool http_get(char *url, UT_string *curl_buf)
 	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writemem);
 	curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)curl_buf);
 
+	fprintf(stderr, "http_get(%s)\n", url);
```
Comment (jpmens) [inline geo.c]: is this left over from debugging?

File: geo.c [hunk 2]
```diff
@@ -212,11 +313,14 @@ int revgeo_getdata(char *apikey, double lat, double lon, UT_string *addr, UT_str
 
 void revgeo_init()
 {
+	fprintf(stderr, "revgeo_init()\n");
```
Comment (jpmens) [inline geo.c]: debugging? please remove

File: geo.c [hunk 3]
```diff
@@ -212,11 +313,14 @@ int revgeo_getdata(char *apikey, double lat, double lon, UT_string *addr, UT_str
...
 	curl = curl_easy_init();
 }
 
 void revgeo_free()
 {
+	fprintf(stderr, "revgeo_free()\n");
```
Comment (jpmens) [inline geo.c]: debugging?

File: lmdb-ll-look.c [hunk 4]
```diff
@@ -25,7 +25,7 @@
 #include "db.h"
 
 #define DBFILE			"data/geocache"
-#define GEOHASH_PRECISION	8
+#define GEOHASH_PRECISION	9
```
Comment (jpmens) [inline lmdb-ll-look.c]: please revert to the original default

File: revgeod.c [hunk 5]
```diff
@@ -49,7 +49,7 @@ statsd_link *sd;
 #endif
 #include "version.h"
 
-#define GEOHASH_PRECISION	8
+#define GEOHASH_PRECISION	9
```
Comment (jpmens) [inline revgeod.c]: please revert to original default

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,16 @@
+FROM debian:bullseye-slim
+
+WORKDIR /usr/src/app
+
+RUN apt-get update \
+     && apt-get install -y --no-install-recommends build-essential jq liblmdb-dev lmdb-utils curl libcurl4-openssl-dev ca-certificates openssl libmicrohttpd12 libmicrohttpd-dev \
```
Comment (jpmens) [inline Dockerfile:6]: is `jq` really required? I don't see where ...

File: README.md [hunk 7]
```diff
@@ -115,7 +115,10 @@ ENVIRONMENT
 if this variable is set when *revgeoc* starts, the program displays received HTTP headers
 
 `OPENCAGE_APIKEY`  
-this mandatory variable must be set in *revgeod*'s environment for it to do reverse geo lookups.
+this optional variable must be set in *revgeod*'s environment for it to do reverse geo lookups via OpenCage.
+
+`LOCATIONIQ_APIKEY`  
+this optional variable must be set in *revgeod*'s environment for it to do reverse geo lookups via [LocationIQ](https://locationiq.com/demo#reverse).
```
Comment (jpmens) [inline README.md:124]: would you please add a sentence to the effect that either OPENCAGE_APIKEY or LOCATIONIQ_APIKEY are required? As it stands they both appear to be optional :)

File: README.md [hunk 8]
```diff
@@ -129,6 +132,15 @@ optionally sets the hostname/address for *revgeoc*; defaults to `127.0.0.1` and
 REQUIREMENTS
 ============
 
+docker
+-----
+Create a .env file with an OPENCAGE_APIKEY or LOCATIONIQ_APIKEY variable defined.
```
Comment (jpmens) [inline README.md:140]: where (in which path) must this .env be created? alongside the docker-compose file? Please say so.

File: docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+#
+version: "3.3"
+ 
+services:
+  revgeod:
+    build:
+      context: ./
+    container_name: revgeod
+    environment:
+      #- OPENCAGE_APIKEY=${OPENCAGE_APIKEY}     # set this in .env
```
Comment (jpmens) [inline docker-compose.yml:17]: "set this in .env" ? Here or there? If either/or: which is better/easier/faster/cheaper? ;)

**Review by jpmens — CHANGES_REQUESTED** [review]
Thank you for your contribution. There are a few nits we'd like addressed.

File: Dockerfile (hunk 6 above)
Comment (halteren) [reply Dockerfile:6]: This is used in the shell scripts (get-stats.sh and c-mini-test.sh). When inside the container it is convenient to have these.

File: docker-compose.yml (hunk 9 above)
Comment (halteren) [reply docker-compose.yml:17]: Since .env is in .gitignore the preferred way is to use .env to avoid accidently checking in your API key(s)

**Comment by jpmens** [comment]
Merged, thank you!

**Pull request merged**